rofl-scheduler/src/serverd/tls.rs

```diff
@@ -0,0 +1,304 @@
+        };
+
+        let (_, cert) = x509_parser::parse_x509_certificate(certificate.end_entity_cert()?)?;
+        match cert.validity().time_to_expiration() {
+            Some(ttl) => {
+                let not_before = cert.validity().not_before;
+                let not_after = cert.validity().not_after;
+                match not_after - not_before {
+                    Some(t) if ttl <= t / 3 => Ok(Duration::ZERO),
+                    _ => Ok(ttl.try_into()?),
+                }
+            }
+            None => Ok(Duration::ZERO),
+        }
+    }
+
+    pub async fn provision_once(&self, acme: &instant_acme::Account) -> Result<()> {
+        slog::info!(self.logger, "provisioning new certificate");
+
+        // Create a new order.
+        let mut order = acme
+            .new_order(&instant_acme::NewOrder::new(&[
+                instant_acme::Identifier::Dns(self.domain.clone()),
+            ]))
+            .await?;
+
+        // Select TLS-ALPN authorization.
+        let mut authorizations = order.authorizations();
+        while let Some(result) = authorizations.next().await {
+            let mut authz = result?;
+            if !matches!(authz.status, instant_acme::AuthorizationStatus::Pending) {
+                continue;
+            }
+
+            let mut challenge = authz
+                .challenge(instant_acme::ChallengeType::TlsAlpn01)
+                .ok_or_else(|| anyhow::anyhow!("no TLS-ALPN01 challenge found"))?;
+
+            // Generate key.
+            let mut params = rcgen::CertificateParams::new(vec![self.domain.clone()])?;
+            params.custom_extensions = vec![rcgen::CustomExtension::new_acme_identifier(
+                challenge.key_authorization().digest().as_ref(),
+            )];
+            let key_pair = KeyPair::generate_for(&PKCS_ECDSA_P256_SHA256)?;
+            let cert = params
+                .self_signed(&key_pair)
+                .context("failed to generate challenge certificate")?;
+            let key_pair = rustls::pki_types::PrivateKeyDer::Pkcs8(
+                rustls::pki_types::PrivatePkcs8KeyDer::from(key_pair.serialize_der()),
+            );
+            let key_pair = rustls::crypto::CryptoProvider::get_default()
+                .ok_or(anyhow!("missing crypto provider"))?
+                .key_provider
+                .load_private_key(key_pair)
+                .context("failed to load challenge private key")?;
+            let certified_key = CertifiedKey::new(vec![cert.der().clone()], key_pair);
+            self.resolver.set_challenge(Some(Arc::new(certified_key)));
+            challenge.set_ready().await?;
+
+            // Currently we only support a single identifier (domain).
+            break;
+        }
+
+        // Exponentially back off until the order becomes ready or invalid.
+        slog::info!(self.logger, "waiting for order to become ready");
+        let status = order.poll(5, Duration::from_millis(250)).await?;
```

Comment:
Is this enough in practice? Looking at the impl, it only waits like ~3.75 secs? But it looks like they only wait for this long in their example as well, so might be enough?

Reply:
It seems to be working fine in practice, I used the recommended defaults, but we can increase the number of retries to `6` for an extra 4 seconds before the final attempt?